### PR TITLE
sync applicationSecret from kubermatic namespace to clusterNamespace

### LINKED ISSUE #10386: sync secret from kubermatic namespace to cluster namespace
To install an application into user-cluster, we need to download the application’s source (e.g. the manifests, helm chart...). To do that, we may need credentials.

Currently, they are stored on master cluster in the kubermatic namespace and replicated on the seed. The application-installation controller belongs to the user-cluster-controller-manager. Therefore it can not access secrets in the kubermatic namespace.

So we to create  have a new controller in the seed-controller-manager that replicates application secrets into all cluster namespaces (e.g `cluster-znx6n7jbdz`)

_note: 
We haven't chosen to add RBAC to provide access to user-cluster-controller-manager to the kubermatic namespace because this namespace has many sensible secrets (especially on master/seed setup)._

### PR DESCRIPTION
**Pull request by vgramer**
**What does this PR do / Why do we need it**:

To install an application into user cluster, we need to download the application’s source (e.g. the manifests, helm chart...). To do that, we may need credentials.

Currently, they are stored on master cluster in the `kubermatic` namespace and replicated on the seed. The application-installation controller belongs to the user-cluster-controller-manager. Therefore it can not access secrets in the `kubermatic` namespace.

This PR adds a new controller in the seed-controller-manager that replicates application secrets into all cluster namespaces 

**Does this PR close any issues?**:
Fixes #10386

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/apps.kubermatic/v1/types.go [hunk 1]
```diff
@@ -23,4 +23,7 @@ const (
 	// ApplicationInstallationCleanupFinalizer indicates that application installed on user-cluster need cleanup
 	// ie uninstall the application, remove  namespace where application were installed ...
 	ApplicationInstallationCleanupFinalizer = "kubermatic.k8c.io/cleanup-application-installation"
+
+	// ApplicationSecretCleanupFinalizer indicates that secret synced from kubermatic namespace to cluster namespace need cleanup.
+	ApplicationSecretCleanupFinalizer = "kubermatic.k8c.io/cleanup-application-secret"
```
Comment (xrstf) [inline pkg/apis/apps.kubermatic/v1/types.go]: Does this need to be a global finalizer or could this simply be a private const in a controller package?

Comment (vgramer) [reply pkg/apis/apps.kubermatic/v1/types.go]: can do in the controller

File: pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, handler.EnqueueRequestsFromMapFunc(enqueueSecret(ctx, r.client, r.namespace)), workerlabel.Predicates(workerName), noDeleteEventPredicate()); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &kubermaticv1.Cluster{}, clusterPauseKey, func(rawObj ctrlruntimeclient.Object) []string {
+		cluster := rawObj.(*kubermaticv1.Cluster)
+		return []string{strconv.FormatBool(cluster.Spec.Pause)}
+	}); err != nil {
+		log.Fatalw("failed to add index on cluster.Spec.Pause: %w", err)
```
Comment (xrstf) [inline pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go]: Please return this error instead.

File: pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForObject{}, predicateutil.ByAnnotation(applicationsecretsynchronizer.SecretTypeAnnotation, "", false), predicateutil.ByNamespace(r.namespace)); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, handler.EnqueueRequestsFromMapFunc(enqueueSecret(ctx, r.client, r.namespace)), workerlabel.Predicates(workerName), noDeleteEventPredicate()); err != nil {
```
Comment (xrstf) [inline pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go:101]: Wouldn't it make sense to also reconcile clusters that are in deletion?

**Review by xrstf — CHANGES_REQUESTED** [review]
Two nit picks :)

File: pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go (hunk 3 above)
Comment (vgramer) [reply pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go:101]: I miss your point...  Application secrets will be deleted at the same time as cluster's namespace. So we can ignore delete events, don't we ?

Comment (xrstf) [reply pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go:101]: You're right.

**Review by xrstf — APPROVED** [review]
/approve

**Pull request merged**